vector2/vector2.go

```diff
@@ -302,116 +374,222 @@ func (v Vector[T]) Add(other Vector[T]) Vector[T] {
 	}
 }
 
+func (v Vector[T]) DivByVector(o Vector[T]) Vector[T] {
+	return Vector[T]{
+		x: v.x / o.x,
+		y: v.y / o.y,
+	}
+}
+
 func (v Vector[T]) DivByConstant(t float64) Vector[T] {
 	return v.Scale(1.0 / t)
 }
 
-func (v Vector[T]) DistanceSquared(other Vector[T]) float64 {
+func (v Vector[T]) Project(normal Vector[T]) Vector[T] {
+	vdn := float64(v.Dot(normal))
+	ndn := float64(normal.Dot(normal))
+	mag := vdn / ndn
+	return normal.Scale(mag)
+}
+
+func (v Vector[T]) Reject(normal Vector[T]) Vector[T] {
+	return v.Sub(v.Project(normal))
+}
+
+func (v Vector[T]) Reflect(normal Vector[T]) Vector[T] {
+	return v.Sub(normal.Scale(2. * float64(v.Dot(normal))))
+}
+
+func (v Vector[T]) DistanceSquared(other Vector[T]) T {
 	xDist := other.x - v.x
 	yDist := other.y - v.y
-	return float64((xDist * xDist) + (yDist * yDist))
+	return (xDist * xDist) + (yDist * yDist)
 }
 
 // Distance is the euclidean distance between two points
 func (v Vector[T]) Distance(other Vector[T]) float64 {
-	return math.Sqrt(v.DistanceSquared(other))
+	return math.Sqrt((float64)(v.DistanceSquared(other)))
 }
 
 // Round takes each component of the vector and rounds it to the nearest whole
 // number
 func (v Vector[T]) Round() Vector[T] {
 	return Vector[T]{
-		x: T(math.Round(float64(v.x))),
-		y: T(math.Round(float64(v.y))),
+		x: mathex.Round(v.x),
+		y: mathex.Round(v.y),
 	}
 }
 
 // RoundToInt takes each component of the vector and rounds it to the nearest
 // whole number, and then casts it to a int
 func (v Vector[T]) RoundToInt() Vector[int] {
 	return New(
-		int(math.Round(float64(v.x))),
-		int(math.Round(float64(v.y))),
+		int(mathex.Round(v.x)),
+		int(mathex.Round(v.y)),
 	)
 }
 
 // Ceil applies the ceil math operation to each component of the vector
 func (v Vector[T]) Ceil() Vector[T] {
 	return Vector[T]{
-		x: T(math.Ceil(float64(v.x))),
-		y: T(math.Ceil(float64(v.y))),
+		x: mathex.Ceil(v.x),
+		y: mathex.Ceil(v.y),
 	}
 }
 
 // CeilToInt applies the ceil math operation to each component of the vector,
 // and then casts it to a int
 func (v Vector[T]) CeilToInt() Vector[int] {
 	return New(
-		int(math.Ceil(float64(v.x))),
-		int(math.Ceil(float64(v.y))),
+		int(mathex.Ceil(v.x)),
+		int(mathex.Ceil(v.y)),
 	)
 }
 
 func (v Vector[T]) Floor() Vector[T] {
 	return Vector[T]{
-		x: T(math.Floor(float64(v.x))),
-		y: T(math.Floor(float64(v.y))),
+		x: mathex.Floor(v.x),
+		y: mathex.Floor(v.y),
 	}
 }
 
 // FloorToInt applies the floor math operation to each component of the vector,
 // and then casts it to a int
 func (v Vector[T]) FloorToInt() Vector[int] {
 	return New(
-		int(math.Floor(float64(v.x))),
-		int(math.Floor(float64(v.y))),
+		int(mathex.Floor(v.x)),
+		int(mathex.Floor(v.y)),
 	)
 }
 
 // Abs applies the Abs math operation to each component of the vector
 func (v Vector[T]) Abs() Vector[T] {
 	return Vector[T]{
-		x: T(math.Abs(float64(v.x))),
-		y: T(math.Abs(float64(v.y))),
+		x: T(mathex.Abs(v.x)),
+		y: T(mathex.Abs(v.y)),
```

Comment:
Is the T needed anymore

```suggestion
		x: mathex.Abs(v.x),
		y: mathex.Abs(v.y),
```